sarif/sarif.go

```diff
@@ -0,0 +1,640 @@
+			Title:           title,
+			Uid:             findingUid,
+		},
+		Message: res.Message.Text,
+		Metadata: &ocsf.Metadata{
+			CorrelationUid: res.CorrelationGuid,
+			EventCode:      ruleID,
+			Labels:         labels,
+			Product: &ocsf.Product{
+				Name: &toolName,
+			},
+			Uid: ruleGuid,
+		},
+
+		Severity:   ptr.Ptr(severityID.String()),
+		SeverityId: severityID,
+		StartTime:  ptr.Ptr(s.clock.Now().Unix()),
+		Status:     ptr.Ptr(s.mapResultKind(res.Kind).String()),
+		StatusId:   ptr.Ptr(s.mapResultKind(res.Kind)),
+		Time:       s.clock.Now().Unix(),
+		TimeDt:     timestamppb.New(s.clock.Now()),
+		TypeName:   ptr.Ptr(typeName[int64(classID)*100+int64(activityID)]),
+		TypeUid:    int64(classID)*100 + int64(activityID),
+		Vulnerabilities: []*ocsf.Vulnerability{
+			{
+				AffectedCode:     affectedCode,
+				AffectedPackages: affectedPackages,
+				Cve:              cve,
+				Cwe:              s.mapCWE(*ruleID),
+				Desc:             &desc,
+				FirstSeenTime:    firstSeenTime,
+				FirstSeenTimeDt:  firstSeenTimeDT,
+				FixAvailable:     fixAvailable,
+				IsFixAvailable:   fixAvailable,
+				LastSeenTime:     lastSeenTime,
+				LastSeenTimeDt:   lastSeenTimeDT,
+				Severity:         ptr.Ptr(severityID.String()),
+				Title:            ptr.Ptr(title),
+				VendorName:       &toolName,
+			},
+		},
+	}, nil
+}
+
+func (s *SarifTransformer) mapProperties(props *sarif.PropertyBag) ([]string, error) {
+	if props == nil {
+		return nil, nil
+	}
+	res := props.Tags
+	if props.AdditionalProperties != nil {
+		extra, err := json.Marshal(props.AdditionalProperties)
+		if err != nil {
+			return nil, err
+		}
+
+		res = append(res, string(extra))
+	}
+	return res, nil
+}
+
+func (s *SarifTransformer) mapResultKind(kind sarif.ResultKind) ocsf.VulnerabilityFinding_StatusId {
+	switch kind {
+	case sarif.ResultKindNotApplicable:
+		return ocsf.VulnerabilityFinding_STATUS_ID_SUPPRESSED // false positive or duplicate or user does not want to see for some reason
+	case sarif.ResultKindFail: // scan has failed findings usually used for some sort of release gating, that's an open finding for all intents and purposes
+	case sarif.ResultKindOpen: // open finding, we need to work with it
+		return ocsf.VulnerabilityFinding_STATUS_ID_NEW
+	case sarif.ResultKindPass: // scan has passed, this finding need not exist
+		return ocsf.VulnerabilityFinding_STATUS_ID_RESOLVED
+	case sarif.ResultKindInformational: // finding is info or under review already, someone is taking care of it or it doesn't need attention
+	case sarif.ResultKindReview:
+		return ocsf.VulnerabilityFinding_STATUS_ID_IN_PROGRESS
+	}
+	// by default, a scanner that doesn't know what status the finding has should always think the status is new unless enrichment says otherwise
+	return ocsf.VulnerabilityFinding_STATUS_ID_NEW
+}
+
+func (s *SarifTransformer) extractCVE(text string) string {
+	re := regexp.MustCompile(`(?i)(?:^|[^A-Za-z0-9-])(CVE-\d{4}-\d{4,7})(?:[^A-Za-z0-9-]|$)`)
```

Comment:
Can we instead compile this in the constructor and handle the error?

It would avoid potential panics :)